### PR TITLE
[CI] Save binaries as artifact 

### PR DESCRIPTION
**Pull request by Disservin**
The idea is to save the produced binaries and make them publicly available to download.
These binaries will also have a slightly different versioning. <br>

Currently it's `Stockfish DDMMYY`, local builds that have git will now use the commit date and commit sha in their version name `Stockfish YYYYMMDD-SHA` otherwise it will fallback to local compilation date and nogit `Stockfish YYYYMMDD-nogit` <br>
For a release version the maintainer has to change the `dev` string to the version number of the release.

Each executable is available to download on its own.

Folder Structure
```
Stockfish Binaries windows-2022
│   Copying.txt
│   AUTHORS
│   Top CPU Contributors.txt
│   stockfish-windows-2022-x86-64-avx2.exe
│
└───src
```

Github reports the uncompressed size so the actual download is half of that. <br>

No functional change

### CODE REVIEW

**Comment by vondele** [comment]
we're now making binaries available without profile-build... not sure that is a good idea. If people start using those, they will be pretty slow. So, my gut feeling is that it is better to make a separate action that is not tied to our current CI workflow, but that we can more easily tweak to get to 'release quality' binaries. In my opinion running this just on master would be enough.

**Comment by vondele** [comment]
so ideally, we can get to a workflow that for each commit to master branch provides production quality binaries somewhere (i.e. our official dev versions), with each release these binaries appear as 'assets' under the release tab, and these binaries are such that the can be copied unmodified to https://stockfishchess.org/download/ ...

**Comment by dav1312** [comment]
> with each release these binaries appear as 'assets' under the release tab, and these binaries are such that the can be copied unmodified to [stockfishchess.org/download](https://stockfishchess.org/download/) ...

Or we can directly use the file links from github and save some bandwidth and storage

**Comment by vondele** [comment]
> Or we can directly use the file links from github and save some bandwidth and storage

our downloads are behind cloudflare, so bandwidth is OK. So far, that has been no issue. But well we could consider. Don't know if github has quota on download volume or so, we might run in that quite easily for releases (say a few 100k downloads, each a few 10MB).

**Comment by Sopel97** [comment]
There is already a automated build tool but it's not hooked up to anything. Not sure if it can be hooked up to actions. https://github.com/official-stockfish/Stockfish/pull/3377

**Comment by Disservin** [comment]
New commit uses a separate workflow for creating the binaries. Also the binaries are now profile guided. Compiling for avx512 etc will be a bit difficult since the runner might not have those.

File: .github/workflows/stockfish.yml
Comment (vondele) [inline .github/workflows/stockfish.yml:14]: this points to Disservin repo?

Comment (Disservin) [reply .github/workflows/stockfish.yml:14]: Yes since I don’t know if it can point to pull request and as of now this file doesn’t exist in stockfish master

Comment (Disservin) [reply .github/workflows/stockfish.yml:14]: This would be something you need to change when a merge is happening

Comment (dav1312) [reply .github/workflows/stockfish.yml:14]: Why don't you simply run the workflow on push to master?

Comment (Disservin) [reply .github/workflows/stockfish.yml:14]: So that you can see how it would look like ?^^ However, now that you have seen it I can change it so that the binaries are only created on push and change the link to official stockfish

**Comment by Disservin** [comment]
The avx2 profile guided compilation for macOS is as of now disabled, somehow profile data was not collected… if someone has any idea please let me know. I can also add profile guided compilation for  Apple silicon

**Comment by Disservin** [comment]
I have changed it so that the binaries only run on master. Of course this CI now does not work because I also linked to the stockfish repo which does not have the file yet 
<br>
`[Invalid workflow file: .github/workflows/stockfish.yml#L14](https://github.com/official-stockfish/Stockfish/actions/runs/3399970200/workflow)
error parsing called workflow "official-stockfish/Stockfish/.github/workflows/stockfish_binaries.yml@master": failed to fetch workflow: workflow was not found.`

File: src/misc.cpp
Comment (vondele) [inline src/misc.cpp:154]: I think we should probably only add date and sha if we have 'dev' in the version string. Otherwise it is unique anyway, and rather keep official releases 'simple'.

**Comment by Disservin** [comment]
Thanks to @dav1312 the compilation step is now a bit simplified and each executable is available on its own.
local builds will be
`Stockfish DDMMYY`

CI builds will be
`Stockfish DDMMYY-SHA`

Release builds will need to change "dev" to version number in `misc.cpp` and then be
`Stockfish version`

**Comment by Sopel97** [comment]
please use ISO dates

**Comment by Disservin** [comment]
> please use ISO dates

DDMMYY is the currently scheme. Kept it in the same format but we might as well change that now I dont think any third party tool relies on stockfishs date format..

**Comment by dav1312** [comment]
Probably worth adding bmi2 to the binaries

**Comment by Disservin** [comment]
> Probably worth adding bmi2 to the binaries

The problem is the runners must have bmi2, not sure if thats the case

**Comment by Disservin** [comment]
Date should be now `YYYYMMDD`

File: .github/workflows/stockfish.yml
Comment (vondele) [inline .github/workflows/stockfish.yml:13]: I'm wondering about this one.. why do we need to reference this? Can the action be triggered independently of the main CI one?

Comment (Disservin) [reply .github/workflows/stockfish.yml:13]: Yes we can however, if you take a look at the finished run https://github.com/Disservin/Stockfish/actions/runs/3400409465 .
We have the normal workflow on the left + a Binaries subdirection for the second workflow. Also the artifacts are at the bottom. If we remove this call then 2 independent workflows will be called. So you have to open two seperate tabs in the browser to view each. Like here (same commit)
https://github.com/Disservin/Stockfish/actions/runs/3400065471
https://github.com/Disservin/Stockfish/actions/runs/3400065473
they are not bundled together, and if you negative to the Actions using this View here 
image

you have to make sure you click on the the correct action that you want to get to (either binary or test ci). 
If you negative via the Actions tab both will have the same name (commit message)
image

so its also really hard to differentiate which is which. 
Whereas with the current logic, you make a commit and theres 1 action thats run which consists of 2 workflows. You click on any and you will find you way to the binary workflow or the test workflow.

Comment (vondele) [reply .github/workflows/stockfish.yml:13]: well, but can we have 'relative paths' like if I fork official-stockfish I'd like my workflow to refer to the (forked) stockfish_binaries.yml at least.

Comment (Disservin) [reply .github/workflows/stockfish.yml:13]: Valid point, it's possible and I force pushed with a relative path.

Comment (dav1312) [reply .github/workflows/stockfish.yml:13]: image

Comment (Disservin) [reply .github/workflows/stockfish.yml:13]: Probably because it was skipped because it only runs on push to master and the variable has not been defined since the if comes before the env, moving it down should fix it I guess

**Comment by Disservin** [comment]
Seems like MSYS2 has some problems today..

File: .github/workflows/stockfish_binaries.yml
Comment (Disservin) [reply .github/workflows/stockfish_binaries.yml]: Changed ;)
it now tries to get the git sha and fails silently if the git command fails `git rev-parse --short HEAD 2>/dev/null`

File: src/misc.cpp
Comment (vondele) [inline src/misc.cpp:151]: now outdated comment.

Comment (Disservin) [reply src/misc.cpp]: fixed

**Comment by ddobbelaere** [comment]
I like the way this is going. Isn't it a good idea to also try to extract the date from git (similar to sha in makefile), but fall back to compile date if not available?

It would have the advantage that the version does not depend on compile date (and e.g. abrok will have the same, or CCC, or...), but reflects date of latest commit.

**Comment by Disservin** [comment]
> I like the way this is going. Isn't it a good idea to also try to extract the date from git (similar to sha in makefile), but fall back to compile date if not available?
> 
> It would have the advantage that the version does not depend on compile date (and e.g. abrok will have the same, or CCC, or...), but reflects date of latest commit.

I think that would obfuscate it quite a bit. Two different dates with different meaning…

**Comment by ddobbelaere** [comment]
> I think that would obfuscate it quite a bit. Two different dates with different meaning…

At the moment not really, as the presence of the git short revision indicates that it indicates the last commit date. The current behavior is (in theory) racy: if different runners get spawned around midnight, artifact version names might be different.

**Comment by Disservin** [comment]
Well I like the idea of using git commit date, but then some compiles which are done in non git repos will have the current date while other will have the commit date (even though it’s date-sha in that case) I think this will create confusion because it’s not so clear to what each date is referring

**Comment by Disservin** [comment]
One automatic way would be to let the runner sign into GitHub stockfish then change the datefile in our repository and commit that. So that whenever you clone or simply download Stockfish it will take the commit date and nothing else… this is something not as easy I think and the Maintainers would have to agree with that…
However even here it can be “racy“ if you clone stockfish before the ci part has finished you still have the new commit but old commit date.

**Comment by ddobbelaere** [comment]
Another way is to add a date file to the archived tarballs (similar to the top CPU contributoers file), and fall back to that. People building the source from this tarball are good then, for all the other ones (e.g. those downloading SF sources as zip from GitHub directly) maybe just output nothing? E.g. only Stockfish dev. Then never any confusion: date always means commit date.

**Comment by Disservin** [comment]
> Another way is to add a date file to the archived tarballs (similar to the top CPU contributoers file), and fall back to that. People building the source from this tarball are good then, for all the other ones (e.g. those downloading SF sources as zip from GitHub directly) maybe just output nothing? E.g. only Stockfish dev. Then never any confusion: date always means commit date.

That’s what I said no? This has other issues though.

**Comment by ddobbelaere** [comment]
> That’s what I said no? This has other issues though.

No, I meant just include it in the tarball but don't commit it. This file should not be versioned....

**Comment by vondele** [comment]
It think it is a good idea to use the git committer date as time stamp, and fall back on the compile date if that doesn't work. The fact that we have a different format (date-sha vs date) clarifies this enough )(we could even use `-nogit` instead of a `-sha` to make it explicit). I wouldn't add files to the repo automatically.

File: src/Makefile
Comment (ddobbelaere) [inline src/Makefile:686]: I think the names `SHA` and `CDATA` or too generic or unclear. Please consider e.g. `GIT(_COMMIT)_SHA/GIT(_COMMIT)_DATE` or `(LAST_)COMMIT_SHA/(LAST_)COMMIT_DATE`.

File: src/misc.cpp
Comment (ddobbelaere) [inline src/misc.cpp:146]: It's [good practice to keep scopes small](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es5-keep-scopes-small), so maybe move to where it is needed (below the `#else`). Then you can also drop the `[[maybe_unused]]` attributes.

**Review by ddobbelaere — CHANGES_REQUESTED** [review]

**Comment by Disservin** [comment]
Okay, it's now using the commit date and otherwise compilation date + nogit

File: src/Makefile
Comment (Disservin) [reply src/Makefile:686]: I tried to keep these short to not litter the Make invocation. So I guess I'll name them `GIT_SHA` and `GIT_DATE`

File: src/misc.cpp
Comment (Disservin) [reply src/misc.cpp:146]: This I can change yeah.

**Comment by Disservin** [comment]
It would be also beneficial to check if the binaries work for the specified OS. 
I have started an action on my own repository with the latest changes, it's currently still running but once its finished please test the binaries ;=) (ETA 25min)

https://github.com/Disservin/Stockfish/actions/runs/3403628835

File: src/Makefile
Comment (ddobbelaere) [reply src/Makefile]: Ok, cool. You forgot one replacement on line 688...

Comment (Disservin) [reply src/Makefile]: Oh shit thanks, overlooked that.

**Comment by vondele** [comment]
Looking at artifacts:
* when I unzip the stockfish-ubuntu-20.04-x86-64-avx2.zip the binary stockfish-ubuntu-20.04-x86-64-avx2 doesn't have execution permissions (so need chmod u+x ). 
* for the version, I would keep 'dev' in the name, that's gone now (like dev-20221106-0bc542ec).
* The unzipping creates the files directly in the place of the zip file, our official downloads unzip to a folder. I think the latter is probably better.

**Comment by dav1312** [comment]
> * when I unzip the stockfish-ubuntu-20.04-x86-64-avx2.zip the binary stockfish-ubuntu-20.04-x86-64-avx2 doesn't have execution permissions (so need chmod u+x ).

This is the same as the current official binaries which are not only labeled as "non official" but they also lack execution permissions

**Comment by Disservin** [comment]
> > * when I unzip the stockfish-ubuntu-20.04-x86-64-avx2.zip the binary stockfish-ubuntu-20.04-x86-64-avx2 doesn't have execution permissions (so need chmod u+x ).
> 
> This is the same as the current official binaries which are not only labeled as "non official" but they also lack execution permissions

When you download from https://stockfishchess.org/download/ the content is in a folder. On abrok it's the same as in the CI.

**Comment by Disservin** [comment]
I think the only way to achieve the folder is too create a folder and mv all files there, seems a bit unnecessary to me..

**Comment by vondele** [comment]
hmm, I just realize that misc.cpp needs obviously to be recompiled for the version output to be correct. If the object is kept from previous compiles, the version date/sha would be wrong. This is no problem for our CI, and it is also what currently happens, but worth pointing out.

**Comment by Disservin** [comment]
> hmm, I just realize that misc.cpp needs obviously to be recompiled for the version output to be correct. If the object is kept from previous compiles, the version date/sha would be wrong. This is no problem for our CI, and it is also what currently happens, but worth pointing out.
> 
> Edit: can be taken care of in the Makefile.

What way do you suppose ? Simplest I can think of right now is to simply rm `misc.o`, what did you have in mind?

**Comment by vondele** [comment]
I've started testing locally with a change like:
```
# Force recompilation to ensure version info is up-to-date
misc.o: FORCE
FORCE:
```

**Comment by vondele** [comment]
I also looked into the permission thing. That seems to be related to this being a zip file. A .tgz would avoid these problems on linux

**Comment by Disservin** [comment]
> I also looked into the permission thing. That seems to be related to this being a zip file. A .tgz would avoid these problems on linux

So I cant change the final download, thats always going to be a zip. We could have a .tgz in the zip but I think this would be just really annoying. The other thing I noticed was that the default unzipping on windows keeps all the files in folder with the same name as the zip, on ubuntu/linux however theres no folder and everything is scattered...

**Comment by vondele** [comment]
right ... that scattering of files is what I referred to earlier.

**Comment by Disservin** [comment]
@vondele Moving everything to one folder and uploading this folder seems to be not possible. The CI only updates that folders content...

File: src/Makefile
Comment (vondele) [inline src/Makefile:965]: FORCE should be added to the list of .PHONY targets

File: src/Makefile
Comment (vondele) [inline src/Makefile:691]: trailing space

File: src/Makefile
Comment (vondele) [inline src/Makefile:685]: trailing space

File: src/misc.cpp
Comment (vondele) [inline src/misc.cpp:157]: I would add a `ss << "dev-"; ` here so the name is shown as `Stockfish dev-20221106-3ed938471 by...`

File: src/misc.cpp
Comment (vondele) [inline src/misc.cpp:143]: This comment here is now outdated

File: src/misc.cpp
Comment (Disservin) [inline src/misc.cpp:158]: @vondele What you want is done here, though I just realised in case of a release version the `-` is superfluous, let me change that.

**Comment by Disservin** [comment]
I think I fixed all reviews, you can have a look again (sorry :p)

**Comment by vondele** [comment]
Will have a look. Meanwhile the permission thing is discussed here https://github.com/actions/upload-artifact#maintaining-file-permissions-and-case-sensitive-files

File: src/misc.cpp
Comment (vondele) [inline src/misc.cpp:156]: trailing spaces.

File: src/misc.cpp
Comment (vondele) [inline src/misc.cpp:159]: Indentation, this block should be aligned with 'version' in the if statement

**Comment by Disservin** [comment]
https://github.com/Disservin/Stockfish/actions/runs/3405055308

Files are now in a tar archive. Also updated checkout from 2 to 3 to silence node js warning.

**Comment by mstembera** [comment]
If we make binaries available would it make sense to make it the ones compiled with clang instead of gcc?  Last time it was tested the clang binaries were faster.

**Comment by vondele** [comment]
I certainly wouldn't make switching to clang part of this PR. gcc is so far the best tested compiler, i.e. dominate fishtest. However, this PR will make such switches rather easy. Other things that could be considered in the future are somehow integrating https://github.com/official-stockfish/Stockfish/pull/3377 or simpler, at least verifying the bench of each generated binary.

**Comment by vondele** [comment]
I've merged this, thanks for working on this...

**Comment by Disservin** [comment]
Also thanks to @dav1312 for your debugging help

**Comment by mstembera** [comment]
@Disservin @vondele  When I build SF on Windows ```make build ARCH=x86-64-bmi2 COMP=mingw -j``` it keeps rebuilding misc.cpp each time I run the command.   As if misc.cpp was modified instead of doing nothing after the first build.  This started with this patch.  Any ideas?  It would be nice to fix before https://github.com/official-stockfish/Stockfish/pull/4250

**Comment by Craftyawesome** [comment]
> ... it keeps rebuilding misc.cpp each time I run the command. As if misc.cpp was modified instead of doing nothing after the first build...

Should happen on linux. And it needs to happen to update the version https://github.com/official-stockfish/Stockfish/pull/4220/files#diff-3e2513390df543315686d7c85bd901ca9256268970032298815d2f893a9f0685R963

**Comment by Disservin** [comment]
> @Disservin @vondele When I build SF on Windows `make build ARCH=x86-64-bmi2 COMP=mingw -j` it keeps rebuilding misc.cpp each time I run the command. As if misc.cpp was modified instead of doing nothing after the first build. This started with this patch. Any ideas? It would be nice to fix before #4250 (I don't know if it happens on Linux also.)

this was actually done on purpose...https://github.com/official-stockfish/Stockfish/pull/4220#issuecomment-1304802206

**Comment by vondele** [comment]
correct, the recompilation is forced to ensure the version reported is correct.

**Comment by mstembera** [comment]
Ok thanks for explaining.

**Comment by mstembera** [comment]
@Disservin @vondele Sorry, actually I'm still confused.  If running a second build did nothing as before no new binary would be produced and the version of the existing (previously built) binary would still have the correct version from when it was built.  So why do we need to produce a new binary when nothing has changed?

**Comment by Disservin** [comment]
In case people compile without stockfish being in a git repository it will use the current date as the version info since we can’t know when date changes we need to recompile. 
If it’s inside a git repository and you pull new changes and these changes only affect search.cpp and you build again the version info string will still have the old commit sha and date.

**Comment by vondele** [comment]
First, let's agree this is not something that would block the release, this is not user visible. Fixes for developers we can do on dev. 
The issue at hand is that compiling into misc.cpp the right version string depends on the git sha of the repo, not just misc.cpp, and if there is no git, depends on the date. The easiest is to recompile misc.cpp always. 
The only alternative I see is to generate a VERSION file at compile time that contains the version string, and only update/touch that VERSION file if the content of the VERSION file changes, with a dependency of misc.cpp on the VERSION file. That would avoid recompilation.

**Comment by mstembera** [comment]
> In case people compile without stockfish being in a git repository it will use the current date as the version info since we can’t know when date changes we need to recompile. 

Sorry for my ignorance but if the binary already exists in the directory why do we need to recompile at all?

>If it’s inside a git repository and you pull new changes and these changes only affect search.cpp and you build again the version info string will still have the old commit sha and date.

Why don't we update the version info string any time any source file changes?


> First, let's agree this is not something that would block the release, this is not user visible. 

Agree.

>The only alternative I see is to generate a VERSION file at compile/make time that contains the version string, and only update/touch that VERSION file if the content of the VERSION file changes, with a dependency of misc.cpp on the VERSION file. That would avoid recompilation.

That would be nice if it's not too much trouble (which I'm not sure).    I think most devs that download SF not familiar w/ this issue would wonder why misc.cpp keeps getting recompiled.

**Comment by Disservin** [comment]
> Sorry for my ignorance but if the binary already exists in the directory why do we need to recompile at all? I mean a binary w/ a newer modified file date than any of the source files.

All good, in that case the date was never really helpful as it always was refering to the compile date. In hindsight the non git one was probably a bad example. The new files from git pull is a much better reasoning why we recompile that file.

Is it really that big of a deal for you that the file gets recompiled ? I mean with parallel make invocation (-j)  it should take about the same time as before I guess since you rarely have to change multiple files. 
I also dont like the VERSION file at compile/make time, just adds unnecessary complexity and you would have to create special rules to listen for file changes, what if you add a new file and/or the file is in gitignore?

**Comment by vondele** [comment]
yes, the version file is not optimal either. Probably adding a dependency of misc.o on all other .o would work. Would need a careful look.

**Comment by mstembera** [comment]
Hmm this is getting deeper down the rabbit hole than I'm able to follow.   I am satisfied that it has been brought to your attention and trust your judgements to decide if it makes sense to change anything.

**Pull request closed without merging**
